sway-fmt-v2/src/items/item_fn.rs

```diff
@@ -70,103 +78,201 @@ impl CurlyBrace for ItemFn {
+                        // `Punctuated<FnArg, CommaToken>`
+                        args.format(formatted_code, formatter)?;
+                    }
                 }
-                // `self`
-                formatted_code.push_str(self_token.span().as_str());
-                // `args_opt`
-                if let Some((comma, args)) = args_opt {
-                    // `, `
-                    write!(formatted_code, "{} ", comma.ident().as_str())?;
-                    // `Punctuated<FnArg, CommaToken>`
-                    args.format(formatted_code, formatter)?;
+                _ => {
+                    format_self(self_token, ref_self, mutable_self, formatted_code)?;
+                    // `args_opt`
+                    if let Some((comma, args)) = args_opt {
+                        // `, `
+                        write!(formatted_code, "{} ", comma.ident().as_str())?;
+                        // `Punctuated<FnArg, CommaToken>`
+                        args.format(formatted_code, formatter)?;
+                    }
                 }
             }
         }
-        // `)`
-        Self::close_parenthesis(formatted_code, formatter)?;
-        // `return_type_opt`
-        if let Some((right_arrow, ty)) = &self.return_type_opt {
-            write!(
-                formatted_code,
-                " {} ",
-                right_arrow.ident().as_str() // `->`
-            )?;
-            ty.format(formatted_code, formatter)?; // `Ty`
-        }
-        // `WhereClause`
-        if let Some(where_clause) = &self.where_clause_opt {
-            writeln!(formatted_code)?;
-            where_clause.format(formatted_code, formatter)?;
-            formatter.shape.update_where_clause();
-        }
-        Ok(())
     }
+
+    Ok(())
+}
+
+fn format_self(
+    self_token: &SelfToken,
+    ref_self: &Option<RefToken>,
+    mutable_self: &Option<MutToken>,
+    formatted_code: &mut FormattedCode,
+) -> Result<(), FormatterError> {
+    // `ref `
+    if let Some(ref_token) = ref_self {
+        write!(formatted_code, "{} ", ref_token.span().as_str())?;
+    }
+    // `mut `
+    if let Some(mut_token) = mutable_self {
+        write!(formatted_code, "{} ", mut_token.span().as_str())?;
+    }
+    // `self`
+    write!(formatted_code, "{}", self_token.span().as_str())?;
+
+    Ok(())
 }
 
-// We will need to add logic to handle the case of long fn arguments, and break into new line
 impl Parenthesis for FnSignature {
     fn open_parenthesis(
         line: &mut FormattedCode,
-        _formatter: &mut Formatter,
+        formatter: &mut Formatter,
     ) -> Result<(), FormatterError> {
-        write!(line, "{}", Delimiter::Parenthesis.as_open_char())?;
+        let open_paren = Delimiter::Parenthesis.as_open_char();
+        match formatter.shape.code_line.line_style {
+            LineStyle::Multiline => {
+                formatter.shape.block_indent(&formatter.config);
+                writeln!(line, "{open_paren}")?;
```

Comment:
Perhaps this line is the culprit of my findings, should this be `write!` instead of `writeln!`?

Reply:
could've swore I got rid of that one, since the tests passed..... weird

Reply:
Ah, I see, my tests all have `self` as the first arg.